modules/stochastic_tools/src/actions/ParameterStudyAction.C

```diff
@@ -0,0 +1,711 @@
+      params.set<SamplerName>("sampler") = samplerName();
+      params.set<std::vector<std::string>>("param_names") =
+          getParam<std::vector<std::string>>("parameters");
+      auto control =
+          _factory.create<Control>("MultiAppSamplerControl", multiappControlName(), params);
+      _problem->getControlWarehouse().addObject(control);
+      if (_show_objects)
+        showObject("MultiAppSamplerControl", multiappControlName(), params);
+    }
+  }
+}
+
+DistributionName
+ParameterStudyAction::distributionName(unsigned int count) const
+{
+  return "study_distribution_" + std::to_string(count);
+}
+
+std::vector<DistributionName>
+ParameterStudyAction::distributionNames(unsigned int full_count) const
+{
+  std::vector<DistributionName> dist_names;
+  for (const auto & i : make_range(full_count))
+    dist_names.push_back(distributionName(i));
+  return dist_names;
+}
+
+ReporterName
+ParameterStudyAction::quantityOfInterestName(const ReporterName & qoi) const
+{
+  return ReporterName(stochasticReporterName(), qoi.getObjectName() + ":" + qoi.getValueName());
+}
+
+void
+ParameterStudyAction::showObject(std::string type,
+                                 std::string name,
+                                 const InputParameters & params) const
+{
+  // Output basic information
+  std::string base_type = params.have_parameter<std::string>("_moose_base")
+                              ? params.get<std::string>("_moose_base")
+                              : "Unknown";
+  _console << "[ParameterStudy] "
+           << "Base Type:  " << COLOR_YELLOW << base_type << COLOR_DEFAULT << "\n"
+           << "                 Type:       " << COLOR_YELLOW << type << COLOR_DEFAULT << "\n"
+           << "                 Name:       " << COLOR_YELLOW << name << COLOR_DEFAULT;
+
+  // Gather parameters and their values if:
+  // - It is not a private parameter
+  // - Doesn't start with '_' (which usually indicates private)
+  // - Parameter is set by this action
+  // - Is not the "type" parameter
+  std::map<std::string, std::string> param_map;
+  for (const auto & it : params)
+    if (!params.isPrivate(it.first) && it.first[0] != '_' && params.isParamSetByUser(it.first) &&
+        it.first != "type")
+    {
+      std::stringstream ss;
+      it.second->print(ss);
+      param_map[it.first] = ss.str();
+    }
+
+  // Print the gathered parameters
+  if (!param_map.empty())
+    _console << "\n                 Parameters: ";
+  bool first = true;
+  for (const auto & it : param_map)
+  {
+    if (!first)
+      _console << "\n" << std::string(29, ' ');
+    _console << COLOR_YELLOW << std::setw(24) << it.first << COLOR_DEFAULT << " : " << COLOR_MAGENTA
+             << it.second << COLOR_DEFAULT;
+    first = false;
+  }
+
+  _console << std::endl;
+}
+
+/**
+ * This class is a hit walker used to see if a list of parameters are all controllable
+ */
+class AreParametersControllableWalker : public hit::Walker
```

Comment:
Why is the declaration of the class in the source file?

Reply:
Moved the declaration to the header file.